File: SMI-SDK/Java SDK/src/pt/sapo/gis/trace/Trace.java
package pt.sapo.gis.trace;

import java.net.InetAddress;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;
import java.util.logging.Level;
import java.util.UUID;
import java.util.logging.Logger;
import pt.sapo.gis.Timer;
import pt.sapo.gis.Event;
import pt.sapo.gis.IEvent;
import pt.sapo.gis.trace.configuration.SeverityType;

/**
 * @author GIS Team
 * Class that implements ITrace interface
 */
public class Trace implements ITrace {

    private final Event<Entry> onEntry = new Event<Entry>();
    private final Event<Trace> onTrace = new Event<Trace>();
    private final List<Entry> entries = new ArrayList<Entry>();
    
    private String id;
    private String contextId;
    private Date start;
    private Date end;
    protected InetAddress serverInfo;
    protected InetAddress clientInfo;        
    protected Timer timer;
    
    //protected Stack<TimerEntry> context = new Stack<TimerEntry>();
    protected static final ThreadLocal<Stack<TimerEntry>> localContext = new ThreadLocal<Stack<TimerEntry>>(){                   
       /* @Override
        protected Stack<TimerEntry> initialValue() {
                return new Stack<TimerEntry>();
             }*/
    };
    
    private Map<String, Object> properties = new HashMap<String, Object>();

    public Map<String, Object> getProperties() {
        return properties;
    }

    public void setProperties(Map<String, Object> properties) {
        this.properties = properties;
    }
    
    public Object getProperty(String key) {
        return properties.get(key);
    }
    
    public void setProperty(String key, Object value) {
        this.properties.put(key, value);
    }

    public InetAddress getClientInfo() {
        return clientInfo;
    }

    public void setClientInfo(InetAddress clientInfo) {
        this.clientInfo = clientInfo;
    }

    public InetAddress getServerInfo() {
        return serverInfo;
    }

    public void setServerInfo(InetAddress serverInfo) {
        this.serverInfo = serverInfo;
    }

    public Date getEnd() {
        return end;
    }

    /**
     * Gets the entries list
     * @return the a read only list of trace entries
     */
    public List<Entry> getEntries() {
        synchronized(entries){
            return Collections.unmodifiableList(entries);
        }
    }   
    
    /**
     * Gets the trace id
     * @return the trace identifier
     */
    public String getId() {
        return id;
    }
    /**
     * Gets the trace id
     * @return the trace context identifier
     */
    public String getContextId() {
        return contextId;
    }
    
    /**
     * Gets the start date
     * @return the trace start date
     */
    public Date getStart() {
        return start;
    }

    /**
     * Gets the size of context
     * @return the number of entries in current context
     */
    public int getContextSize(){
        return localContext.get().size();
    }
    
    /**
     * Creates a new trace     
     */
    public Trace() {
        this(UUID.randomUUID().toString(), null);
    }

    /**
     * Creates a new trace     
     * @param id the trace identifier
     * @param contextId the trace context identifier
     */
    public Trace(String id, String contextId) {
        if (id == null) {
            id = UUID.randomUUID().toString();
        }
        this.id = id;
        this.contextId = contextId;
        this.start = new Date();
        this.timer = new Timer().Start();   
        localContext.set(new Stack<TimerEntry>());
    }

    /**
     * Trace a new entry
     * @param entry the entry to add to the trace
     */
    @Override
    public void trace(Entry entry) {
        entry.setParent(this);
        entry.offset = this.timer.ElapsedMilis();
        addEntry(entry);
    }
    
    private void addEntry(Entry entry){
      if (localContext.get().empty()) {
            synchronized(entries){
                entries.add(entry);
            }
        } else {
            localContext.get().peek().entry.entries.add(entry);
        }
        try {
            onEntry.notify(entry);
        } catch (Exception ex) {
            Logger.getLogger(Trace.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    /**
     * Start a new entry context. Every entry added after this will be
     * added as sub entry of this one.     
     */
    @Override
    public Entry beginEntry() {        
        return beginEntry(new Entry());        
    }

    /**
     * Start a new entry context. Every entry added after this will be
     * added as sub entry of this one.     
     * @param description the new entry description
     * @param severity the new entry severity type
     */
    @Override
    public Entry beginEntry(String description, SeverityType severity) {
        return beginEntry(new Entry(description, severity));
        
    }

    /**
     * Start a new entry context. Every entry added after this will be
     * added as sub entry of this one.     
     * @param description the new entry description
     * @param severity the new entry severity type
     * @param properties the new entry properties map
     */
    @Override
    public Entry beginEntry(String description, SeverityType severity, Map<String, Object> properties) {
        return beginEntry(new Entry(description, severity, properties));        
    }

    /**
     * Start a new entry context. Every entry added after this will be
     * added as sub entry of this one.     
     * @param entry the entry to start the new context     
     */
    @Override
    public Entry beginEntry(Entry entry) {        
        Timer entryTimer = new Timer().Start();
        entry.setParent(this);
        entry.offset = entryTimer.ElapsedMilis();
        localContext.get().push(new TimerEntry(entryTimer, entry));
        return entry;
    }
    
    /**
     * Ends the current entry context, adding it to the trace entries list 
     * or to the parent entry if it was created inside a context     
     */
    @Override
    public void endEntry() {
        TimerEntry current = localContext.get().pop();
        double milis = current.timer.Stop().ElapsedMilis();
        current.entry.duration = milis;
        addEntry(current.entry);
    }

    /**
     * Gets OnEntry event
     * @return the object that implements IEvent of an Entry
     */
    public IEvent<Entry> getOnEntryEvent() {
        return this.onEntry;
    }

    /**
     * Gets OnEntry event
     * @return the object that implements IEvent of an Trace
     */
    public IEvent<Trace> getOnTraceEvent() {
        return this.onTrace;
    }
    /**
     * Finishes the trace closing all the open context entries     
     */
    @Override
    public void finish() {
        this.timer.Stop();
        this.end = new Date();
        while (!localContext.get().empty()) {
            this.endEntry();
        }
        try {
            onTrace.notify(this);
        } catch (Exception ex) {
            Logger.getLogger(Trace.class.getName()).log(Level.SEVERE, null, ex);
        }
        finally{
            localContext.remove();
        }
    }

    /**
     * Gets all severity types that exists on trace entries, including sub entries
     * @return the severity types collection set
     */
    @Override
    public Set<SeverityType> getSeverityTypes() {
        Set<SeverityType> severities = new HashSet<SeverityType>();
        synchronized(entries){
            for (Entry e : this.entries) {
                GetEntrySeverityTypes(e, severities);
            }
        }
        return severities;
    }

    private void GetEntrySeverityTypes(Entry entry, Set<SeverityType> severities) {
        severities.add(entry.getSeverity());
        for (Entry e : entry.getEntries()) {
            GetEntrySeverityTypes(e, severities);
        }
    }
}

File: SMI-SDK/.NET SDK/Library/gis/trace/Trace.cs
namespace pt.sapo.gis.trace
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    using System.Net;
    using pt.sapo.gis.trace.configuration;
    using System.Diagnostics;
    using System.Collections;
    using System.Threading;

    /// <summary>
    /// Type that represents a trace
    /// </summary>
    public class Trace : ITrace
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        private Stopwatch stopwatch = Stopwatch.StartNew();

        public event GenericEventHandler<Entry> OnEntry;
        public event GenericEventHandler<Trace> OnTrace;

        public Guid Id { get; protected set; }
        public String ContextId { get; protected set; }
        public DateTime Start { get; protected set; }
        public DateTime End { get; protected set; }
        public List<Entry> Entries { get; protected set; }
        protected Stack<Entry> Context
        {
            get
            {
                if (context.IsValueCreated == false)
                    context.Value = new Stack<Entry>();
                return context.Value;
            }
        }
        protected ThreadLocal<Stack<Entry>> context = new ThreadLocal<Stack<Entry>>();

        public IPAddress ServerInfo { get; set; }
        public IPAddress ClientInfo { get; set; }
        public Summary Summary { get; set; }
        public IDictionary<String, Object> Properties { get; protected set; }

        /// <summary> 
        /// Creates a new trace     
        /// </summary>
        public Trace()
            : this(Guid.NewGuid(), null) { }

        /// <summary> 
        /// Creates a new trace with a given identifier and context identifier
        /// </summary>
        /// <param name="param name="id">Trace identifier</param>
        /// <param name="contextId">Trace context identifier</param>
        public Trace(Guid id, String contextId)
        {
            if (id == null)
            {
                id = Guid.NewGuid();
            }
            this.Id = id;
            this.ContextId = contextId;
            this.Start = DateTime.UtcNow;
            Entries = new List<Entry>();
            Properties = new Dictionary<String, Object>();
            stopwatch.Start();
        }

        /// <summary>
        /// Trace a new entry
        /// </summary>
        /// <param name="entry">entry the entry to add to the trace</param>
        public void TraceEntry(Entry entry)
        {
            entry.Parent = this;
            entry.Offset = stopwatch.ElapsedMilliseconds;
            AddEntry(entry);
        }

        private void AddEntry(Entry entry)
        {
            if (Context.Count == 0)
            {
                lock (Entries)
                {
                    Entries.Add(entry);
                }
            }
            else
            {
                Context.Peek().Entries.Add(entry);
            }
            try
            {
                if (OnEntry != null)
                {
                    lock (Entries)
                    {
                        OnEntry(entry);
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error(null, ex);
            }
        }

        /// <summary>Start a new entry context. Every entry added after this will be added as sub entry of this one.</summary>
        public Entry BeginEntry()
        {
            return BeginEntry(new Entry());
        }

        /// <summary>Start a new entry context. Every entry added after this will be added as sub entry of this one.</summary>
        /// <param name="description">the new entry description</param>
        /// <param name="severity">the new entry severity type</param>

        public Entry BeginEntry(String description, severityType severity)
        {
            return BeginEntry(new Entry(description, severity));
        }

        /// <summary>Start a new entry context. Every entry added after this will be added as sub entry of this one.</summary>
        /// <param name="description">the new entry description</param>
        /// <param name="severity">the new entry severity type</param>
        /// <param name="properties">the new entry properties map</param>
        public Entry BeginEntry(String description, severityType severity, IDictionary<String, Object> properties)
        {
            return BeginEntry(new Entry(description, severity, properties));
        }

        ///
        /// <summary>Start a new entry context. Every entry added after this will be added as sub entry of this one.</summary>
        /// <param name="entry">the entry to start the new context     </param>
        public Entry BeginEntry(Entry entry)
        {
            entry.Offset = stopwatch.ElapsedMilliseconds;
            entry.Parent = this;
            Context.Push(entry);
            return entry;
        }

        /// <summary>Ends the current entry context, adding it to the trace entries list or to the parent entry if it was created inside a context</summary>
        public void EndEntry()
        {
            var current = Context.Pop();
            current.Duration = stopwatch.ElapsedMilliseconds - current.Offset;
            AddEntry(current);
        }

        /// <summary>Finishes the trace closing all the open context entries</summary>
        public void Finish()
        {
            End = DateTime.UtcNow;
            while (Context.Count > 0)
            {
                this.EndEntry();
            }
            try
            {
                if (OnTrace != null)
                    OnTrace(this);
            }
            catch (Exception ex)
            {
                log.Fatal(null, ex);
            }
        }

        /// <summary>Gets all severity types that exists on trace entries, including sub entries</summary>
        /// <returns>the severity types collection set</returns>
        public ISet<severityType> GetSeverityTypes()
        {
            var severities = new HashSet<severityType>();
            lock (Entries)
            {
                foreach (Entry e in this.Entries)
                {
                    GetEntrySeverityTypes(e, severities);
                }
            }
            return severities;
        }

        private void GetEntrySeverityTypes(Entry entry, ISet<severityType> severities)
        {
            severities.Add(entry.Severity);
            foreach (Entry e in entry.Entries)
            {
                GetEntrySeverityTypes(e, severities);
            }
        }
    }
}